packages/server-wallet/src/engine/engine.ts

```diff
@@ -199,6 +199,15 @@ export class SingleThreadedEngine implements EngineInterface {
 
       const {participants} = channel;
       response.queueSendObjective(o, channel.myIndex, participants);
+
+      // Sync any ledger channels used for an open channel objective
+      if (
+        o.type === 'OpenChannel' &&
+        o.data.fundingStrategy === 'Ledger' &&
+        !!o.data.fundingLedgerChannelId
+      ) {
+        await this._syncChannel(o.data.fundingLedgerChannelId, response);
+      }
```

Comment:
This sort of feels like the wrong way to go about it:
- if one channel is out of sync, is there a possibility that many channels are out of sync?
- if there are many channels out of sync, then we would be syncing the ledger channel many times, right?

In my mind, it seems like you'd be better off syncing from the ledger channel "downwards", rather than from the funded channel "upwards".

All in all, I find it hard to keep everything together in my head about what could go wrong, how to detect things going wrong, and how to fix things when things go wrong. This seems like a good fix for the short term, and we should schedule ourselves to come back to the problem during cooldown?

Reply:
> This sort of feels like the wrong way to go about it:
> 
> * if one channel is out of sync, is there a possibility that many channels are out of sync?
> * if there are many channels out of sync, then we would be syncing the ledger channel many times, right?


> In my mind, it seems like you'd be better off syncing from the ledger channel "downwards", rather than from the funded channel "upwards".

Is there a reason to do this besides performance? Right now our model is based on syncing objectives. After an objective does not make progress for a certain amount of time we resend a message containing the objective and channel states. So it feels natural to include the ledger information in the "sync objective" payload, as it's needed to make progress on the objective.

To sync from a ledger channel "downwards" I think we'd need to change our approach to poll ledger channels and check when they were last updated.

If performance is an issue it's pretty trivial to fetch all the ledger channels at once for a batch of objectives we want to sync.
  
> All in all, I find it hard to keep everything together in my head about what could go wrong, how to detect things going wrong, and how to fix things when things go wrong. This seems like a good fix for the short term, and we should schedule ourselves to come back to the problem during cooldown?

Could you elaborate on what you mean by "go wrong"?  It seems like in the worst case we may send messages including ledger information already known by another participant, but that doesn't seem too bad?

Reply:
> > This sort of feels like the wrong way to go about it:
> > 
> > * if one channel is out of sync, is there a possibility that many channels are out of sync?
> > * if there are many channels out of sync, then we would be syncing the ledger channel many times, right?
> 
> > In my mind, it seems like you'd be better off syncing from the ledger channel "downwards", rather than from the funded channel "upwards".
> 
> Is there a reason to do this besides performance? Right now our model is based on syncing objectives. After an objective does not make progress for a certain amount of time we resend a message containing the objective and channel states. So it feels natural to include the ledger information in the "sync objective" payload, as it's needed to make progress on the objective.

At the same time, the ledger information is relevant to _many_ objectives, and the ledger protocol is quite complicated, so I (for one) can't really understand what would happen if we synced many objectives at once, which were all open channel objectives funded by the same ledger channel. This change is introducing some potentially overwhelming behaviour into the wallet, and that is what I'm expressing concern about.

> Does syncing from a ledger channel "downwards" mean polling the ledger database and making sure those ledger channels make progress?

There is a ledger funding protocol that I do not believe anyone on the team really understands. Therefore, I don't think that anybody can really say for sure. As a general idea, though, since the ledger protocol is not objective-driven, and is designed to handle concurrent ledger requests, it feel like it should not be synced when syncing an objective for a child channel.

> > All in all, I find it hard to keep everything together in my head about what could go wrong, how to detect things going wrong, and how to fix things when things go wrong. This seems like a good fix for the short term, and we should schedule ourselves to come back to the problem during cooldown?
> 
> Could you elaborate on what you mean by "go wrong"?

"go wrong" is pretty vague, but there are a few behaviours that we _didn't think would happen_ that we then _observed did happen_, either in some production environment or as a result of running the new test infrastructure (or even the existing e2e tests, as shown by the flickering test that we ignored!)

In my opinion, we'd be in better shape if we started with a good argument of why things _shouldn't_ "go wrong"; as this PR points out, we don't have such an argument (things "went wrong" in a very simple situation) and I am advocating that we think about it and come up with one.

> It seems like in the worst case we may send messages including ledger information already known by another participant, but that doesn't seem too bad?

- it depends on how many messages we send. If we're sending the same message 100 times, and it takes some real amount of time or resources to process that message (as I can imagine being true for a well-used ledger channel) that could be a problem. Is it possible to "go critical" and trigger more failures?
- is syncing a ledger channel enough to ensure that ledger funding proceeds? why? what about ledger requests that never made it into a state update? does syncing a ledger channel also sync the ledger requests?
- what about closing ledger funded channels?
- what about opening/closing virtually funded channels?

What I'm trying to get at is, I don't know a good argument why the proposed solution in this PR is a good solution long term. At a minimum, it does not cover things that can go wrong today (messages dropped while closing a ledger-funded channel), and at worst, it could lead to unpredictable overloading of a wallet due to redundant messages.

Reply:
@andrewgordstewart I've created an issue for this, would you be ok merging this as-is for now? (And if so could you approve it?)

Reply:
done!